tests/src/test/java/org/eclipse/hono/tests/amqp/AmqpAdapterBase.java

```diff
@@ -202,88 +175,110 @@ public void testUploadingXnumberOfMessages(final TestContext context) throws Int
-                .setHandler(context.asyncAssertSuccess(ok -> addTenantTracker.complete()));
-        addTenantTracker.await();
-
-        final Async setup = context.async();
-
-        final HonoClient client = prepareAmqpAdapterClient(username, password);
-        client.connect(new ProtonClientOptions()).compose(ok -> {
-            clientsToShutdown.add(client);
-            return Future.succeededFuture();
-        }).compose(ok -> createConsumer(tenantId, msg -> {
-            received.countDown();
-            if (received.getCount() % 40 == 0) {
-                LOG.info("message {} received", MESSAGES_TO_SEND - received.getCount());
-            }
-        })).setHandler(context.asyncAssertSuccess(ok -> setup.complete()));
-        setup.await();
-
-        final AtomicInteger messageCount = new AtomicInteger(0);
-
-        while (messageCount.get() < MESSAGES_TO_SEND) {
+        doUploadMessages(context, tenantId, payload -> {
             final Async sendingComplete = context.async();
-            getOrCreateSender(tenantId, client).compose(sender -> {
-                final Message message = ProtonHelper.message("temp: " + messageCount.getAndIncrement());
-                sender.send(message).setHandler(sendOutcome -> {
-                    if (sendOutcome.succeeded()) {
-                        if (messageCount.get() % 40 == 0) {
-                            LOG.info("messages sent: {}", messageCount.get());
-                        }
-                    } else {
-                        LOG.debug("error sending message {}", messageCount.get(), sendOutcome.cause());
-                    }
-                    sendingComplete.complete();
-                });
-
-                return Future.succeededFuture();
+            final Message msg = ProtonHelper.message(payload);
+            sender.send(msg).setHandler(outcome -> {
+                sendingComplete.complete();
             });
             sendingComplete.await();
-        }
+        });
+    }
+
+    // ------------------------------< private methods >---
+
+    /**
+     * Sets up the protocol adapter by doing the following:
+     * <ul>
+     * <li>Add a device (with credentials) for the tenant identified by the given tenantId.</li>
+     * <li>Create an AMQP 1.0 client and authenticate it to the protocol adapter with username: {@code [device-id@tenantId]}.</li>
+     * <li>After a successful connection, create a producer/sender for sending messages to the protocol adapter.</li>
+     * </ul>
+     *
+     * @param tenantId The tenant to register with the device registry.
+     * @param deviceId The device to add to the tenant identified by tenantId.
+     * @param disableTenant If true, disable the protocol adapter for the tenant.
+     * @param ctx The Vert.x test context.
+     */
+    private void setupProtocolAdapter(final String tenantId, final String deviceId, final boolean disableTenant, final TestContext ctx) {
```

Comment:
why not return a `Future<MessageSender>`?

Reply:
ok sure..

Reply:
humm.. just see that this will make the tests less more readable :-)